contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,210 @@
+        return _revokeRoles(resource, roleBitmap, account);
+    }
+
+    // Internal functions
+
+    /**
+     * @dev Reverts if `account` does not have all the given roles.
+     */
+    function _checkRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        if (!hasRoles(resource, roleBitmap, account)) {
+            revert EACUnauthorizedAccountRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Reverts if `account` does not have the admin roles for all the given roles.
+     */
+    function _checkCanGrantRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        uint256 settableRoles = _getSettableRoles(resource, account);
+        if ((roleBitmap & ~settableRoles) != 0) {
+            revert EACUnauthorizedAccountAdminRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Copies all roles from `srcAccount` to `dstAccount` within the same resource.
+     *
+     * @param resource The resource to copy roles within.
+     * @param srcAccount The account to copy roles from.
+     * @param dstAccount The account to copy roles to.
+     */
+    function _copyRoles(bytes32 resource, address srcAccount, address dstAccount) internal virtual {
+        uint256 srcRoles = _roles[resource][srcAccount];
+        _roles[resource][dstAccount] |= srcRoles;
+        emit EACRolesGranted(resource, srcRoles, dstAccount);
+    }
+
+    /**
+     * @dev Grants multiple roles to `account`.
+     */
+    function _grantRoles(bytes32 resource, uint256 roleBitmap, address account) internal virtual returns (bool) {
+        uint256 currentRoles = _roles[resource][account];
+        uint256 updatedRoles = currentRoles | roleBitmap;
+
+        if (currentRoles != updatedRoles) {
+            _roles[resource][account] = updatedRoles;
+            emit EACRolesGranted(resource, roleBitmap, account);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @dev Attempts to revoke roles from `account` and returns a boolean indicating if roles were revoked.
+     */
+    function _revokeRoles(bytes32 resource, uint256 roleBitmap, address account) internal virtual returns (bool) {
+        uint256 currentRoles = _roles[resource][account];
+        uint256 updatedRoles = currentRoles & ~roleBitmap;
+        
+        if (currentRoles != updatedRoles) {
+            _roles[resource][account] = updatedRoles;
+            emit EACRolesRevoked(resource, roleBitmap, account);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @dev Revoke all roles for account within resource.
+     */
+    function _revokeAllRoles(bytes32 resource, address account) internal virtual returns (bool) {
+        return _revokeRoles(resource, _roles[resource][account], account);
```

Comment:
This can avoid a storage read by just specifying `0xff..ff` for the `roleBitmap`.